AP_TECS: scale velRateMin correctly with airspeed

coleschon opened the pull request:
This PR improves TECS deceleration rates by allowing to decrease speed more rapidly, the faster a plane is going.

### Description
The current calculation for `velRateMin` scales incorrectly with airspeed, where currently, the max allowable velocity rate _decreases_ as velocity increases. i.e. the faster an aircraft is going, the less-quickly it's able to slow down. This is particularly problematic on aircraft which operate over a wide range of speeds or which have a large difference between `TECS_SINK_MIN` and `TECS_SINK_MAX`, since they are empirically determined.

This change introduces discrete calculations for:
* `_STEdot_neg_max` — Specific total energy rate at max speed (using `TECS_SINK_MAX`)
* `velRateNegCruise` — Velocity rate at cruise speed (existing velocity rate calc at `TRIM_ARSPD_CM`)
* `velRateNegMax` — Velocity rate at max speed (existing velocity rate calc at `ARSPD_FBW_MAX`)

The change then replaces the `velRateMin` calculation with a linear interpolation between the velocity rates at cruise and max speed, which is capped at those values outside the bounds of cruise and max speed. The deceleration margin is also decreased from 50% to 10% to improve deceleration rate performance.

### Testing
This change was tested in ArduPilot SITL ([Plane-4.3.3](https://github.com/ArduPilot/ardupilot/tree/Plane-4.3.3)) with the following sequence:
1. Takeoff the plane
2. Send guided mode command to a distant point or for a large loiter
3. Issue `MAV_CMD_DO_CHANGE_SPEED` commands to increase and decrease speed setpoint
4. Issue `MAV_CMD_DO_CHANGE_ALTITUDE` commands to increase and decrease altitude setpoint

#### Before
Notice the blue at line at the top of graph (behind the green) representing `TECS.sp`, this is the plane's airspeed. Before this change, it required just over 6 minutes for the plane to decelerate from 75 m/s to 40 m/s. Extremely slow, and not by limitation of the aircraft  — throttle is being modulated to maintain a gradual deceleration.

![Current TECS example](https://github.com/ArduPilot/ardupilot/assets/24783918/2cf7a257-488d-46e4-ad25-b0c4cac09228)

#### After
Notice again the blue line representing `TECS.sp`. After this change, the plane requires only 40 seconds to decelerate from 75 m/s to 40 m/s.

![Updated TECS example](https://github.com/ArduPilot/ardupilot/assets/24783918/d0c247c6-ad80-4b52-a1da-d9a9e1bab127)

_Note:_ all TECS parameters are identical between runs.

#### Summary
Here we can see the rates produced from our capped linear interpolation between velocity rate at cruise and max speeds compared against what currently exists.


![velocity vs airspeed](https://github.com/ArduPilot/ardupilot/assets/24783918/048534ee-0e0d-4cdd-9470-62585aca37a0)

**X 40** m/s = `TRIM_ARSPD_CM` (cruise airspeed)
**X 80** m/s = `ARSPD_FBW_MAX` (max airspeed)

coleschon commented:
What is the repo's guidance for backporting improvements like this? My team is currently running `Plane-4.3.3` — would we have to own a fork with the changes applied on top of this branch? I'd love to get this merged into `master` in either case.

IamPete1 commented on libraries/AP_TECS/AP_TECS.cpp:
```diff
@@ -470,11 +470,20 @@ void AP_TECS::_update_speed_demand(void)
     _TAS_dem = constrain_float(_TAS_dem, _TASmin, _TASmax);
 
+    // Determine the true cruising airspeed (m/s)
+    _TAScruise = 0.01f * (float)aparm.airspeed_cruise_cm * _ahrs.get_EAS2TAS();
```
```suggestion
    const float TAScruise = 0.01f * (float)aparm.airspeed_cruise_cm * _ahrs.get_EAS2TAS();
```
I think this can just be local?

coleschon replied:
Yes, since we're not using it elsewhere.

IamPete1 commented on libraries/AP_TECS/AP_TECS.h:
```diff
@@ -266,6 +266,9 @@ class AP_TECS {
     float _TASmax;
     float _TASmin;
 
+    // True cruising airspeed (TRIM_ARSPD)
+    float _TAScruise;
+
```
```suggestion
```

IamPete1 commented on libraries/AP_TECS/AP_TECS.cpp:
```diff
@@ -470,11 +470,20 @@ void AP_TECS::_update_speed_demand(void)
+    // Maximum permissible rate of deceleration value at cruise speed
+    const float velRateNegCruise = 0.9f * _STEdot_min / _TAScruise;
+    // Linear interpolation between velocity rate at cruise and max speeds
+    const float velRateFinal = (((velRateNegMax - velRateNegCruise) / (_TASmax - _TAScruise)) * _TAS_state) + (velRateNegCruise - ((velRateNegMax - velRateNegCruise) / (_TASmax - _TAScruise)) * _TAScruise);
+    const float velRateMin = constrain_float(velRateFinal, velRateNegMax, velRateNegCruise);
```
```suggestion
const float velRateMin = linear_interpolate(velRateNegMax, velRateNegCruise, _TAS_state, _TASmax, TAScruise);
```
I think you can just use our build in linear interpolate function?

coleschon replied:
Yup, good call.
```cpp
const float velRateMin = constrain_float(linear_interpolate(velRateNegMax, velRateNegCruise, _TAS_state, _TASmax, TAScruise), velRateNegMax, velRateNegCruise);
```

IamPete1 replied:
The `linear_interpolate` function already does the constrain.

coleschon replied:
Updated.

IamPete1 commented:
@coleschon You explain the testing that was done, but not issue that this fixes. What are the symptoms and  how does the fix improve those? 

RE backporting, we would get it into master first. I don't think this is a critical fix it would not be back-ported but it would be included in the next stable release, 4.5. If you need it on 4.3 you would have to maintain your own branch.

Hwurzburg reviewed (changes requested):
squash and proper commit message needed

tridge commented:
@priseborough can you check this?

coleschon commented:
@IamPete1 I agree that this should use the built-in `linear_interpolate()` function and am testing now to verify it behaves as expected. Also, I added a before/after example in **Testing** to illustrate the issues with the current deceleration rates. Please let me know if you think this is sufficient.

@Hwurzburg squashed commits and corrected message.

coleschon commented:
@priseborough @IamPete1 @Hwurzburg ready for re-review.

IamPete1 commented:
@priseborough Can you take a look at this please, Thanks.

IamPete1 commented:
@coleschon Sorry we were so slow getting this in, Thanks for sticking with us.
